text/0000-native-class-constructor-update.md

```diff
@@ -0,0 +1,235 @@
+- Start Date: 2018-06-14
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Native Class Constructor Update
+
+## Summary
+
+Update the behavior of EmberObject's constructor to defer object
+initialization.
+
+## Motivation
+
+Using native class syntax with EmberObject has almost reached full feature
+parity, meaning soon we'll be able to ship native classes and begin recommending
+them. This will do wonders for the Ember learning story, and will bring us in
+line with the wider Javascript community.
+
+However, early adopters of native classes have experienced some serious
+ergonomic issues due to the current behavior of the class constructor. The issue
+is caused by the fact that properties passed to `EmberObject.create` are
+assigned to the instance in the root class `constructor`. Due to the way that
+native class fields work, this means that they are assigned _before_ any
+subclasses' fields are assigned, causing subclass fields to overwrite any value
+passed to `create`:
+
+```js
+class Foo extends EmberObject {
+  bar = 'baz';
+}
+
+let foo = Foo.create({ bar: 'something different' });
+
+console.log(foo.bar); // 'baz'
+```
+
+This has made adoption very difficult, and is a consistent stumbling block for
+new users of native class syntax in Ember. Worse yet, it makes writing a codemod
+for converting to native class syntax very difficult because we don't have a
+clear target.
+
+For instance, given the above class, how would we convert the class field? Let's
+go through the various options:
+
+```js
+class Foo extends EmberObject {
+  // Does not work, for the reasons described above
+  bar = 'baz';
+
+  // Does not cover all cases. If we did `Foo.create({ bar: false })` it would
+  // still assign the default.
+  bar = this.bar || 'baz';
+
+  // This works, but is very verbose and not ideal
+  bar = this.hasOwnProperty('bar') ? this.bar : 'baz';
+
+  // This is one of the community accepted solutions, but it requires lodash
+  bar = _.defaultTo(this.bar, 'baz');
+
+  // This is another community accepted solution, but it requires
+  // @ember-decorators/argument, which is a separate library
+  @argument foo = 'bar';
+}
+```
+
+None of these is ideal. Instead, we can change the behavior of the constructor
+and the `create` method to circumvent this issue.
+
+This change _would_ be a breaking change to the behavior of native classes
+today, and a change from the previous class RFC. This will impact early adopters
+and should be made with that in mind. It would _not_ be a change that breaks the
+behavior of the community solutions to class fields mentioned above, and all
+other changes would be relatively easy to create a safe codemod for (essentially
+converting `constructor` -> `init` in affected classes), so the impact _should_
+be minimal.
+
```

Comment:
An alternative not mentioned here is to make the change behind a feature flag enabled by an add-on. This would make the new behavior opt in. We are doing this already for making jQuery optional, for example.

Reply:
That would be a reasonable alternative, we should definitely add it to the alternatives section.

In practice, I think it would be best to ship the target behavior as the actual/only public API and polyfill its behavior to older versions of Ember (already looking into how we can do that, it _should_ be possible). Unlike a lot of the other optional feature flags, we don't _technically_ have a public API that we need to continue to support.

This does beg the question, when does an RFC spec become public API? The process in intentionally fuzzy here, because details may change in implementation (see Named Blocks).

In this case, we've talked to lots of early adopters and they have so far been very supportive of this late-stage update to the original RFC. However, if there are any users of native class syntax out there who would be very against it, we definitely want to hear from them!